src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -16,74 +16,69 @@
 import com.google.gson.Gson;
 import com.google.sps.data.User;
+import java.util.Map;
+import java.util.HashMap;
 
 @WebServlet("/login")
 public class LoginServlet extends HttpServlet {
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-
     UserService userService = UserServiceFactory.getUserService();
 
     String urlToRedirect = "/";
     String userId = "null";
     String logoutUrl = "null";
     String loginUrl = "null";
     String userEmail = "null";
-
-    // Boolean to update User's main page 
     Boolean showNewTab = false;
 
+    User user;
     if (userService.isUserLoggedIn()) {
-      userEmail = userService.getCurrentUser().getEmail();
       userId = userService.getCurrentUser().getUserId();
+      userEmail = userService.getCurrentUser().getEmail();
       logoutUrl = userService.createLogoutURL(urlToRedirect);
       showNewTab = true;
     } else {
       loginUrl = userService.createLoginURL(urlToRedirect);
     }
 
-    User user = new User(userId, loginUrl, logoutUrl, userEmail, showNewTab);
-    
-    if (userId != "null") {
-        checkUserInDatastore(user);
+    user = new User(userId, userEmail);
+
+    if (userId != "null" && !isUserInDatastore(user)) {
+      storeUserToDatastore(user);
     }
 
+    Map<String, Object> jsonInfo = new HashMap<>();
+    jsonInfo.put("showTabStatus", showNewTab);
+    jsonInfo.put("logoutUrl", logoutUrl);
+    jsonInfo.put("loginUrl", loginUrl);
+
     response.setContentType("application/json;");
-    response.getWriter().println(new Gson().toJson(user));
+    response.getWriter().println(new Gson().toJson(jsonInfo));
   }
 
-  public void checkUserInDatastore(User user) {
-    // ** NEED TO WRITE TEST FOR BEHAVIORAL TESTING
-
-    Key userKey = KeyFactory.createKey("User", user.getUserId());
+  public boolean isUserInDatastore(User user) {
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    Entity userEntity;
+    Key userKey = KeyFactory.createKey("User", user.getUserId());
 
-    // Check if user is in datastore
     try {
-      userEntity = datastore.get(userKey);
+      datastore.get(userKey);
       user.setUserKey(KeyFactory.keyToString(userKey));
+      return true;
     } catch (EntityNotFoundException e) {
-      userEntity = null;
-    }
-
-    if (userEntity == null) {
-        addUserToDatastore(datastore, user);
+      return false;
     }
   }
 
-  public void addUserToDatastore(DatastoreService datastore, User user) {
+  public void storeUserToDatastore(User user) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
-    // Create new user entity 
     Entity newEntity = user.createEntity();
-
-    // Create User's key and add to object
-    Key key = newEntity.getKey(); 
+    Key key = newEntity.getKey();
     String userKeyStr = KeyFactory.keyToString(key);
     user.setUserKey(userKeyStr);
```

Comment:
Same here as above. Again, doesn't have to be this PR, but this method should return the key and let the calling code change the user.

Keeping methods pure has two benefits.

1) The method is more reusable, because sometimes you don't want/need those side effects.
2) It's easier to reason about pure functions, because you don't have to be worried that they are changing state.